From 823079e853b45c8009345eb0cea7fd5adef889f0 Mon Sep 17 00:00:00 2001
From: morckx <morckx@gmail.com>
Date: Sat, 20 Dec 2014 22:43:39 +0100
Subject: [PATCH] MountService: allow overriding default storage list with
 ro.storage_list.override

Useful for devices with different storage layout variants, e.g. with
either emulated or non-emulated sdcard0.

Change-Id: Ia44ef6243563a69cb8d38a7b87bd81cd37f1df63
---
 .../core/java/com/android/server/MountService.java    | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/services/core/java/com/android/server/MountService.java b/services/core/java/com/android/server/MountService.java
index 0b04bb0..f24c01c 100755
--- a/services/core/java/com/android/server/MountService.java
+++ b/services/core/java/com/android/server/MountService.java
@@ -1366,6 +1366,25 @@ private void readStorageListLocked() {
         Resources resources = mContext.getResources();
 
         int id = com.android.internal.R.xml.storage_list;
+
+        /*
+         * If ro.storage_list.override is set use it to override the
+         * default. This is used for devices with different storage
+         * layouts, e.g. with either emulated or non-emulated sdcard0.
+         */
+        String storageListOverride = SystemProperties.get("ro.storage_list.override");
+        if (!storageListOverride.isEmpty()) {
+            int tmp_id = resources.getIdentifier(storageListOverride, "xml",
+                                                 mContext.getPackageName());
+            if(tmp_id > 0) {
+                Slog.i(TAG, "readStorageListLocked: using storage list "
+                       + storageListOverride);
+                id = tmp_id;
+            } else {
+                Slog.e(TAG, "readStorageListLocked: could not retrieve storage list "
+                       + storageListOverride + " using default instead");
+            }
+        }
         XmlResourceParser parser = resources.getXml(id);
         AttributeSet attrs = Xml.asAttributeSet(parser);
 
